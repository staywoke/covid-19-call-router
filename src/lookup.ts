import zipcodes from 'zipcodes';
import statesCsv from './data/states.csv';
import caZipCodeCountiesCsv from './data/ca-zipcode-counties.csv';
import caCountryNumbersCsv from './data/ca-county-numbers.csv';

export interface State {
  name: string;
  code: string;
  phoneNumber?: string;
}

export interface County {
  name: string;
  phoneNumber?: string;
}

export interface Zip {
  code: string;
  state: State;
  county?: County;
  getCovidPhoneNumber: () => string | null;
}

const CALIFORNIA: State = {
  code: 'CA',
  name: 'California',
};

export const lookupZip = (code: string): Zip | null => {
  // Find location of zip
  const location = zipcodes.lookup(code);

  if (!location) {
    return null;
  }

  if (location.state === CALIFORNIA.code) {
    return (
      caZips.find(caZip => {
        const pattern = new RegExp(caZip.code);
        return pattern.test(code);
      }) || null
    );
  }

  const state = states.find(
    state => state.code === location.state,
  ) as State;

  return {
    code,
    state,
    getCovidPhoneNumber: () => {
      return state.phoneNumber as string;
    },
  };
};

export const normalizePhoneNumber = (
  phoneNumber: string,
): string | undefined => {
  return phoneNumber
    ? '+1' + phoneNumber.replace(/[^0-9]/g, '')
    : undefined;
};

export const parseCsv = <T>(
  csv: string,
  mapRow: (row: string[]) => T,
): T[] => {
  return csv
    .trim()
    .split('\n')
    .map(line => mapRow(line.trim().split('\t')));
};

export const caCounties = parseCsv<County>(
  caCountryNumbersCsv,
  ([name, phoneNumber]): County => ({
    name: name.toUpperCase(),
    phoneNumber: normalizePhoneNumber(phoneNumber),
  }),
);

export const caZips = parseCsv<Zip>(
  caZipCodeCountiesCsv,
  ([code, countyName]): Zip => {
    const county = caCounties.find(county => {
      return county.name === countyName;
    });

    return {
      code,
      county,
      state: CALIFORNIA,
      getCovidPhoneNumber: () => {
        return county?.phoneNumber || null;
      },
    };
  },
);

export const states = parseCsv<State>(
  statesCsv,
  ([name, code, phoneNumber]) => ({
    name,
    code,
    phoneNumber: normalizePhoneNumber(phoneNumber),
  }),
);
