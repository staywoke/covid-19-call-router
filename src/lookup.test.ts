import { lookupZip } from './lookup';
import validZipCodes from './data/valid-zips';

describe('get phone number by zip', () => {
  test('non-CA zip returns zip object without county, state w/ phone number', () => {
    const zip = lookupZip('11201');
    expect(zip?.state.code).toEqual('NY');
    expect(zip).toMatchSnapshot();
    expect(zip?.getCovidPhoneNumber()).toEqual(
      zip?.state.phoneNumber,
    );
  });

  test('CA zip returns zip object with county, state w/o phone number', () => {
    const zip = lookupZip('92811');
    expect(zip).toMatchSnapshot();
    expect(zip?.getCovidPhoneNumber()).toEqual(
      zip?.county?.phoneNumber,
    );
  });
});

test('all valid zip codes are matched', () => {
  const bad: string[] = [];

  for (const zipCode of validZipCodes) {
    const zip = lookupZip(zipCode);

    if (zip === null) {
      bad.push(zipCode);
    }
  }

  throw new Error(
    `These zip codes did not return results:\n${bad.join('\n')}`,
  );
});
