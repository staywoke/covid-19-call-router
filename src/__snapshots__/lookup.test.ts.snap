// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get phone number by zip CA zip returns zip object with county, state w/o phone number 1`] = `
Object {
  "code": "928[0-9]{2}",
  "county": Object {
    "name": "ORANGE",
    "phoneNumber": "+18886004357",
  },
  "getCovidPhoneNumber": [Function],
  "state": Object {
    "code": "CA",
    "name": "California",
  },
}
`;

exports[`get phone number by zip non-CA zip returns zip object without county, state w/ phone number 1`] = `
Object {
  "code": "11201",
  "getCovidPhoneNumber": [Function],
  "state": Object {
    "code": "NY",
    "name": "New York",
    "phoneNumber": "+18883643065",
  },
}
`;
